R/fread.R

```diff
@@ -36,6 +36,33 @@ yaml=FALSE, autostart=NA, tmpdir=tempdir(), tz="UTC")
   )
   nThread=as.integer(nThread)
   stopifnot(nThread>=1L)
+
+  is_url = function(x) any(startsWith(x, c("http://", "https://", "ftp://", "ftps://", "file://")))
+
+  download_file = function(input) {
+    str7 = substr(input, 1L, 7L) # avoid grepl() for #2531
+    # nocov start
+    if (str7=="ftps://" || startsWith(input, "https://")) {
+      if (!requireNamespace("curl", quietly = TRUE))
+        stopf("Input URL requires https:// connection for which fread() requires 'curl' package which cannot be found. Please install 'curl' using 'install.packages('curl')'.") # nocov
+      tmpFile = tempfile(fileext = paste0(".",tools::file_ext(input)), tmpdir=tmpdir)  # retain .gz extension in temp filename so it knows to be decompressed further below
+      curl::curl_download(input, tmpFile, mode="wb", quiet = !showProgress)
+    }
+    else { #if (startsWith(input, "ftp://") || str7== "http://" || str7=="file://") {
+      method = if (str7=="file://") "internal" else getOption("download.file.method", default="auto")
+      # force "auto" when file:// to ensure we don't use an invalid option (e.g. wget), #1668
+      tmpFile = tempfile(fileext = paste0(".",tools::file_ext(input)), tmpdir=tmpdir)
+      download.file(input, tmpFile, method=method, mode="wb", quiet=!showProgress)
+      # In text mode on Windows-only, R doubles up \r to make \r\r\n line endings. mode="wb" avoids that. See ?connections:"CRLF"
+    }
+    assign("file", tmpFile, envir=parent.env(environment()))
```

Comment:
Why not just return `tmpFile`?

Reply:
Maybe my view was locked in thinking about the different environments. A return is definitely the more straightforward way.